rct_optimizations/test/dh_chain_kinematic_measurement_utest.cpp

```diff
@@ -0,0 +1,352 @@
+    std::cout << "Orientation difference Std. Dev.: " << std::sqrt(ba::variance(ori_acc)) << std::endl;
+
+    EXPECT_LT(ba::mean(pos_acc), position_diff_mean_threshold);
+    EXPECT_LT(std::sqrt(ba::variance(pos_acc)), position_diff_std_dev_threshold);
+    EXPECT_LT(ba::mean(ori_acc), orientation_diff_mean_threshold);
+    EXPECT_LT(std::sqrt(ba::variance(ori_acc)), orientation_diff_std_dev_threshold);
+  }
+
+  double residual_threshold_sqr;
+
+  double position_diff_mean_threshold;
+  double position_diff_std_dev_threshold;
+  double orientation_diff_mean_threshold;
+  double orientation_diff_std_dev_threshold;
+
+  std::size_t n_observations;
+
+  KinematicMeasurement::Set observations;
+
+  const DHChain camera_chain;
+  const DHChain target_chain;
+  Eigen::Isometry3d camera_mount_to_camera;
+  Eigen::Isometry3d target_mount_to_target;
+  Eigen::Isometry3d camera_base_to_target_base;
+
+  KinematicCalibrationProblemPose3D problem;
+};
+
+/**
+ * @brief Tests the Dual DH Chain kinematic calibration algorithm with perfect initial conditions
+ */
+class DHChainMeasurementTest_PerfectInitial : public DHChainMeasurementTest
+{
+public:
+  using DHChainMeasurementTest::DHChainMeasurementTest;
+
+  virtual void analyzeResults(const KinematicCalibrationResult& result) override
+  {
+    // Check the nominal expectations
+    DHChainMeasurementTest::analyzeResults(result);
+
+    EXPECT_LT(result.initial_cost_per_obs, 1.0e-15);
+
+    EXPECT_TRUE(result.camera_mount_to_camera.isApprox(camera_mount_to_camera));
+    EXPECT_TRUE(result.target_mount_to_target.isApprox(target_mount_to_target));
+    EXPECT_TRUE(result.camera_base_to_target_base.isApprox(camera_base_to_target_base));
+
+    // Expect the resulting DH parameter offsets to be the same shape as the input
+    // This is an important check of the case where the cost function modifies for 0 DoF chains
+    EXPECT_EQ(problem.camera_chain.dof(), result.camera_chain_dh_offsets.rows());
+    EXPECT_EQ(problem.target_chain.dof(), result.target_chain_dh_offsets.rows());
+
+    // Expect the camera DH offsets to remain the same as the input (i.e. all zeros)
+    for (Eigen::Index row = 0; row < result.camera_chain_dh_offsets.rows(); ++row)
+    {
+      for (Eigen::Index col = 0; col < result.camera_chain_dh_offsets.cols(); ++col)
+      {
+        EXPECT_LT(result.camera_chain_dh_offsets(row, col), 1.0e-15);
+      }
+    }
+  }
+};
+
+/**
+ * @brief Tests the Dual DH Chain kinematic calibration algorithm with
+ * initial guesses for DH params that are slightly different from the true values
+ */
+class DHChainMeasurementTest_PerturbedDH : public DHChainMeasurementTest
+{
+public:
+  using DHChainMeasurementTest::DHChainMeasurementTest;
+
+  virtual void setActualData() override
+  {
+    DHChainMeasurementTest::setActualData();
+
+    problem.camera_chain = test::createABBIRB2400();
+
+    // Set the residual squared error threshold to 0.5 pixels
+    residual_threshold_sqr = 0.5 * 0.5;
```

Comment:
What is the residual error here? Is it some mixture of units since we're calculating the difference between position and orientation?